databuilder/dummy_data/query_info.py

```diff
@@ -0,0 +1,128 @@
+    """
+
+    name: str
+    type: type  # NOQA: A003
+    categories: Optional[tuple]
+    values_used: set = dataclasses.field(default_factory=set)
+
+    def __post_init__(self):
+        if hasattr(self.type, "_primitive_type"):
+            self.type = self.type._primitive_type()
+
+    def record_value(self, value):
+        if hasattr(value, "_to_primitive_type"):
+            value = value._to_primitive_type()
+        self.values_used.add(value)
+
+
+@dataclasses.dataclass
+class TableInfo:
+    """
+    Captures information about a table as used in a particular dataset definition
+    """
+
+    name: str
+    has_one_row_per_patient: bool
+    columns: dict[str, ColumnInfo] = dataclasses.field(default_factory=dict)
+
+
+@dataclasses.dataclass
+class QueryInfo:
+    """
+    Captures information about a dataset definition, including the specific tables and
+    columns used and the values expected in those columns
+    """
+
+    tables: dict[str, TableInfo]
+
+    @classmethod
+    def from_variable_definitions(cls, variable_definitions):
+        all_nodes = walk_tree(*variable_definitions.values())
+        by_type = get_nodes_by_type(all_nodes)
+
+        tables = {
+            # Create a TableInfo object …
+            table.name: TableInfo(
+                name=table.name,
+                has_one_row_per_patient=isinstance(table, SelectPatientTable),
+            )
+            # … for every table used in the query (sorted for consistency)
+            for table in sort_by_name(
+                by_type[SelectTable] | by_type[SelectPatientTable]
+            )
+        }
+
+        column_info_by_column: dict[SelectColumn, ColumnInfo] = {}
+
+        # For every column used in the query (sorted for consistency) …
+        for column in sort_by_name(by_type[SelectColumn]):
+            table = get_root_frame(column.source)
+            name = column.name
+            table_info = tables[table.name]
+            column_info = table_info.columns.get(name)
+            if column_info is None:
+                # … insert a ColumnInfo object into the appropriate table
+                column_info = ColumnInfo(
+                    name=name,
+                    type=table.schema.get_column_type(name),
+                    categories=table.schema.get_column_categories(name),
+                )
+                table_info.columns[name] = column_info
+            # Record the ColumnInfo object associated with each SelectColumn node
+            column_info_by_column[column] = column_info
+
+        # Record values used in equality and substring comparisons
+        for node in by_type[Function.EQ] | by_type[Function.StringContains]:
+            # The query model in theory supports "1 == x" style comparisons (i.e.  with
+            # the column on the RHS) but there's no way to generate such constructions
+            # using ehrQL so we only bother handling the "x == 1" orientation here.
```

Comment:
What if something in ehrQL changes, or if we use another QM front end?

Reply:
We'll fail to record some values and generate "worse quality" dummy data. Under some circumstances that could mean that we can't generate any matching patients at all. But given all the other ways dummy data generation can go wrong it didn't strike me as hugely high priority. And it's a pain to write tests for because you have to construct the QM by hand.

Reply:
Fair enough!